crates/sui-package-resolver/src/lib.rs

```diff
@@ -356,6 +359,96 @@ impl<S: PackageStore> Resolver<S> {
 
         Ok(sigs)
     }
+
+    /// Attempts to infer the type layouts for pure inputs to the programmable transaction.
+    ///
+    /// The returned vector contains an element for each input to `tx`. Elements corresponding to
+    /// pure inputs that are used as arguments to transaction commands will contain `Some(layout)`.
+    /// Elements for other inputs (non-pure inputs, and unused pure inputs) will be `None`.
+    ///
+    /// Layout resolution can fail if a type/module/package doesn't exist, if layout resolution hits
+    /// a limit, or if a pure input is somehow used in multiple conflicting occasions (with
+    /// different types).
+    pub async fn pure_input_layouts(
+        &self,
+        tx: &ProgrammableTransaction,
+    ) -> Result<Vec<Option<MoveTypeLayout>>> {
+        let mut tags = vec![None; tx.inputs.len()];
+        let mut register_type = |arg: &Argument, tag: &TypeTag| {
+            let &Argument::Input(ix) = arg else {
+                return Ok(());
+            };
+
+            if !matches!(tx.inputs.get(ix as usize), Some(CallArg::Pure(_))) {
+                return Ok(());
+            }
+
+            let Some(type_) = tags.get_mut(ix as usize) else {
+                return Ok(());
+            };
+
+            if let Some(prev) = type_.replace(tag.clone()) {
+                // SAFETY: We just inserted `tag` in here.
+                let curr = type_.take().unwrap();
+                return Err(Error::InputTypeConflict(ix, prev, curr));
+            };
+
+            Ok(())
+        };
+
+        // (1). Infer type tags for pure inputs from their uses.
+        for cmd in &tx.commands {
+            match cmd {
+                Command::MoveCall(call) => {
+                    let params = self
+                        .function_parameters(
+                            call.package.into(),
+                            call.module.as_str(),
+                            call.function.as_str(),
+                        )
+                        .await?;
+
+                    for (open_sig, arg) in params.iter().zip(call.arguments.iter()) {
+                        let sig = open_sig.instantiate(&call.type_arguments)?;
```

Comment:
The use of proper type instantiation is an enhancement on the previous implementation, which ignored type parameters. This means that in a situation like this:

```
public fun foo<T>(xs: vector<T>);
```

It couldn't detect from a call to `foo<u8>` that its parameter was a byte array. This implementation detects such cases.